sdk/core/azure-core/src/main/java/com/azure/core/util/SharedExecutorService.java

```diff
@@ -0,0 +1,259 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util;
+
+import com.azure.core.implementation.ReflectionUtils;
+import com.azure.core.implementation.ReflectiveInvoker;
+import com.azure.core.util.logging.ClientLogger;
+
+import java.time.Duration;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * An {@link ExecutorService} that is shared by multiple consumers.
+ * <p>
+ * On creation of this executor service, it will register a runtime shutdown hook and reference shutdown hook to ensure
+ * it is cleaned up when either the JVM exits or is no longer being referenced.
+ * <p>
+ * This executor service may be created by using its constructor, but it is recommended to use the shared instance.
+ */
+public final class SharedExecutorService implements ExecutorService {
+    private static final ClientLogger LOGGER = new ClientLogger(SharedExecutorService.class);
+
+    private static final AtomicReference<SharedExecutorService> INSTANCE = new AtomicReference<>();
+
+    // Shared thread counter for all instances of SharedExecutorService created using the empty factory method.
+    private static final AtomicLong AZURE_SDK_THREAD_COUNTER = new AtomicLong();
+    private static final String AZURE_SDK_THREAD_NAME = "azure-sdk-global-thread-";
+
+    // The thread pool size for the shared executor service.
+    //
+    // This uses the configuration setting 'azure.sdk.threadPoolSize' if set, otherwise it defaults to 10 times the
+    // number of available processors.
+    // If 'azure.sdk.threadPoolSize' is set to a non-integer, negative value, or zero, the default value is used.
+    private static final int THREAD_POOL_SIZE
+        = Configuration.getGlobalConfiguration().get("azure.sdk.threadPoolSize", config -> {
+            try {
+                int size = Integer.parseInt(config);
+                if (size <= 0) {
+                    return 10 * Runtime.getRuntime().availableProcessors();
+                } else {
+                    return size;
+                }
+            } catch (NumberFormatException ignored) {
+                return 10 * Runtime.getRuntime().availableProcessors();
+            }
+        });
```

Comment:
Thinking a long way ahead - if we port this to generic core, what would the config parameter be, and do we want the ability to look up multiple keys to reduce the code duplication?